### PR TITLE
Allow setting custom protocol in mplex

### LINKED ISSUE #2344: [Feature Request] Allow setting custom protocol in mplex
in [go-libp2p-muxer](https://github.com/libp2p/go-libp2p/blob/master/options.go#L101), it allows using custom protocol name. Adding a similar pub fn set_protocol to MplexConfig to connect to go nodes with custom protocol.  

```go
// Muxer configures libp2p to use the given stream multiplexer (or stream
// multiplexer constructor).
//
// Name is the protocol name.
//
// The transport can be a constructed mux.Transport or a function taking any
// subset of this libp2p node's:
// * Peer ID
// * Host
// * Network
// * Peerstore
func Muxer(name string, tpt interface{}) Option {
	mtpt, err := config.MuxerConstructor(tpt)
	err = traceError(err, 1)
	return func(cfg *Config) error {
		if err != nil {
			return err
		}
		cfg.Muxers = append(cfg.Muxers, config.MsMuxC{MuxC: mtpt, ID: name})
		return nil
	}
}
```

### PR DESCRIPTION
**Pull request by hanabi1224**
in [go-libp2p-muxer](https://github.com/libp2p/go-libp2p/blob/master/options.go#L101), it allows using custom protocol name.  Adding a similar `pub fn set_protocol` to `MplexConfig`  to connect to go nodes with custom protocol

```go
// Muxer configures libp2p to use the given stream multiplexer (or stream
// multiplexer constructor).
//
// Name is the protocol name.
//
// The transport can be a constructed mux.Transport or a function taking any
// subset of this libp2p node's:
// * Peer ID
// * Host
// * Network
// * Peerstore
func Muxer(name string, tpt interface{}) Option {
	mtpt, err := config.MuxerConstructor(tpt)
	err = traceError(err, 1)
	return func(cfg *Config) error {
		if err != nil {
			return err
		}
		cfg.Muxers = append(cfg.Muxers, config.MsMuxC{MuxC: mtpt, ID: name})
		return nil
	}
}
```

Closes https://github.com/libp2p/rust-libp2p/issues/2344

### CODE REVIEW

File: muxers/mplex/src/config.rs [hunk 1]
```diff
@@ -84,6 +88,17 @@ impl MplexConfig {
         self.split_send_size = size;
         self
     }
+
+    /// Gets protocol
+    pub fn protocol(&self) -> &[u8] {
+        self.protocol
+    }
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]: None of the other fields have a _getter_ method. Why is this one needed?

File: muxers/mplex/src/config.rs [hunk 2]
```diff
@@ -21,6 +21,8 @@
 use crate::codec::MAX_FRAME_SIZE;
 use std::cmp;
 
+pub(crate) const DEFAULT_MPLEX_PROTOCOL: &'static [u8] = b"/mplex/6.7.0";
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]:
```suggestion
pub(crate) const DEFAULT_MPLEX_PROTOCOL_NAME: &'static [u8] = b"/mplex/6.7.0";
```

How about referring to this as the protocol **name**?

File: muxers/mplex/src/config.rs [hunk 3]
```diff
@@ -84,6 +88,17 @@ impl MplexConfig {
...
         self
     }
+
+    /// Gets protocol
+    pub fn protocol(&self) -> &[u8] {
+        self.protocol
+    }
+
+    /// Sets protocol, muxer.set_protocol(b"/mplex/6.7.0")
+    pub fn set_protocol(&mut self, protocol: &'static [u8]) -> &mut Self {
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]:
```suggestion
    pub fn set_protocol_name(&mut self, protocol: &'static [u8]) -> &mut Self {
```

File: muxers/mplex/src/config.rs [hunk 4]
```diff
@@ -84,6 +88,17 @@ impl MplexConfig {
         self.split_send_size = size;
         self
     }
+
+    /// Gets protocol
+    pub fn protocol(&self) -> &[u8] {
+        self.protocol
+    }
+
+    /// Sets protocol, muxer.set_protocol(b"/mplex/6.7.0")
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]:
```suggestion
    /// Set protocol name.
    ///
    /// ```rust
    /// muxer_config.set_protocol(b"/mplex/6.7.0");
    /// ```
```

File: muxers/mplex/src/config.rs [hunk 5]
```diff
@@ -120,6 +135,7 @@ impl Default for MplexConfig {
             max_buffer_len: 32,
             max_buffer_behaviour: MaxBufferBehaviour::Block,
             split_send_size: 8 * 1024,
+            protocol: DEFAULT_MPLEX_PROTOCOL,
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]:
```suggestion
            protocol_name: DEFAULT_MPLEX_PROTOCOL,
```

**Review by mxinden — COMMENTED** [review]
Could you include a changelog entry and bump the version?

https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md

File: muxers/mplex/src/config.rs (hunk 1 above)
Comment (hanabi1224) [reply muxers/mplex/src/config.rs]: I believe it's useful for debugging/logging purposes to be able to retrieve the protocol name when it can be customized

**Comment by hanabi1224** [comment]
@mxinden Thanks for your review, I've updated the PR, please take another look.

File: muxers/mplex/src/config.rs [hunk 6]
```diff
@@ -84,6 +88,23 @@ impl MplexConfig {
         self.split_send_size = size;
         self
     }
+
+    /// Gets the protocol anme
+    pub fn protocol_name(&self) -> &[u8] {
+        self.protocol_name
+    }
```
Comment (mxinden) [inline muxers/mplex/src/config.rs]: Sorry for the bringing this up once more, still, for the sake of consistency, I don't think we should introduce a getter for a single field only.

In case you would like to introduce getters for all fields, I think that should be a separate pull request.

File: muxers/mplex/Cargo.toml [hunk 7]
```diff
@@ -2,7 +2,7 @@
 name = "libp2p-mplex"
 edition = "2018"
 description = "Mplex multiplexing protocol for libp2p"
-version = "0.30.0"
+version = "0.31.0"
```
Comment (mxinden) [inline muxers/mplex/Cargo.toml]:
```suggestion
version = "0.30.1"
```

Is this a breaking change? If not, why not release it as a patch release?

File: muxers/mplex/CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,8 @@
+# unreleased
```
Comment (mxinden) [inline muxers/mplex/CHANGELOG.md]:
```suggestion
# 0.30.1 [unreleased]
```

**Comment by hanabi1224** [comment]
@mxinden  Comments resolved, plz take another look thx!

File: muxers/mplex/src/config.rs (hunk 6 above)
Comment (hanabi1224) [reply muxers/mplex/src/config.rs]: Removed as suggested. Although I don't agree with the point of maintaining consistency. I think it makes sense to have immutable getters only for a selective set of the fields, with the same rationale behind having different access modifiers for different fields in many languages including rust, but I do agree this might belong to a separate PR, hope this addresses ur concern in this PR, thx!

Comment (hanabi1224) [reply muxers/mplex/src/config.rs]: Removed as suggested. 

Although I don't really agree with the point of maintaining consistency. I think it makes sense to have immutable getters only for a subset of the fields, sharing the same rationale behind having different access modifiers for different fields in many languages including rust, but I do agree this might belong to a separate PR.

Hope this addresses ur concern in this PR, thx!

File: muxers/mplex/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,7 @@
+# 0.30.1
```
Comment (mxinden) [inline muxers/mplex/CHANGELOG.md]:
```suggestion
# 0.30.1 [unreleased]
```

**Review by mxinden — APPROVED** [review]
Thanks @hanabi1224.

**Comment by hanabi1224** [comment]
Thanks!
